packages/flutter/lib/src/widgets/basic.dart

```diff
@@ -2315,50 +2316,156 @@ class ConstrainedBox extends SingleChildRenderObjectWidget {
+/// given.
+///
+/// ```dart
+/// Container(
+///   constraints: BoxConstraints(minHeight: 40, maxHeight: 100),
+///   alignment: Alignment.center,
+///   child: ConstraintsTransformBox(
+///     constraintsTransform: ConstraintsTransformBox.maxHeightUnconstrained,
+///     child: const Card(child: Text('Hello World!')),
+///   )
+/// )
+/// ```
+/// {@end-tool}
+///
 /// See also:
 ///
-///  * [ConstrainedBox], for a box which imposes constraints on its child.
-///  * [Align], which loosens the constraints given to the child rather than
-///    removing them entirely.
-///  * [Container], a convenience widget that combines common painting,
-///    positioning, and sizing widgets.
-///  * [OverflowBox], a widget that imposes different constraints on its child
-///    than it gets from its parent, possibly allowing the child to overflow
-///    the parent.
-class UnconstrainedBox extends SingleChildRenderObjectWidget {
-  /// Creates a widget that imposes no constraints on its child, allowing it to
-  /// render at its "natural" size. If the child overflows the parents
-  /// constraints, a warning will be given in debug mode.
-  const UnconstrainedBox({
+///  * [ConstrainedBox], which renders a box which imposes constraints
+///    on its child.
+///  * [OverflowBox], a widget that imposes additional constraints on its child,
+///    and allows the child to overflow itself.
+///  * [UnconstrainedBox] which allows its children to render themselves
+///    unconstrained, expands to fit them, and considers overflow to be an error.
+class ConstraintsTransformBox extends SingleChildRenderObjectWidget {
+  /// Creates a widget that uses a function to transform the constraints it
+  /// passes to its child. If the child overflows the parent's constraints, a
+  /// warning will be given in debug mode.
+  ///
+  /// The `debugTransformType` argument adds a debug label to this widget.
+  ///
+  /// The `alignment`, `clipBehavior` and `constraintsTransform` arguments must
+  /// not be null.
+  const ConstraintsTransformBox({
     Key? key,
     Widget? child,
     this.textDirection,
     this.alignment = Alignment.center,
-    this.constrainedAxis,
+    required this.constraintsTransform,
     this.clipBehavior = Clip.none,
-  }) : assert(alignment != null),
+    String debugTransformType = '',
+  }) : _debugTransformLabel = debugTransformType,
+       assert(alignment != null),
        assert(clipBehavior != null),
+       assert(constraintsTransform != null),
+       assert(debugTransformType != null),
        super(key: key, child: child);
 
+  /// A [BoxConstraintsTransform] that always returns its argument as-is (i.e.,
+  /// it is an identity function).
+  ///
+  /// The [ConstraintsTransformBox] becomes a proxy widget that has no effect on
+  /// layout if [constraintsTransform] is set to this.
+  static BoxConstraints unmodified(BoxConstraints constraints) => const BoxConstraints();
```

Comment:
Should this return the argument "constraints"? Otherwise its not matching the documentation of being an identity function...